main_test.go

```diff
@@ -1,62 +1,202 @@
 package main
 
 import (
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
 	"testing"
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/assert"
 )
 
-func TestMain(t *testing.T) {
-	RegisterFailHandler(Fail)
-	RunSpecs(t, "main")
+type SSOOptions struct {
+	URL    string
+	Secret string
+	Nonce  string
+	Groups string
+	Admin  bool
 }
 
-var _ = Describe("parseCookie", func() {
-	var (
-		signed, secret string
+type SSOOverrideFunc func(*SSOOptions)
+type ConfigOverrideFunc func(*Config)
 
-		parsedUsername string
-		parsedGroup    string
-		parseError     error
-	)
+func mustParseURL(s string) *url.URL {
+	u, err := url.Parse(s)
+	if err != nil {
+		panic(err)
+	}
+	return u
+}
 
-	JustBeforeEach(func() {
-		parsedUsername, parsedGroup, parseError = parseCookie(signed, secret)
-	})
+func NewTestConfig() Config {
+	originURL := mustParseURL("http://origin.url")
+	proxyURL := mustParseURL("http://proxy.url")
+	ssoURL := mustParseURL("http://sso.url")
 
-	Context("when verifying with an invalid secret", func() {
-		BeforeEach(func() {
-			secret = "secretbar"
-			signed = signCookie("user,group", "secretfoo")
-		})
+	return Config{
```

Comment:
No need for the intermediate variables; the field names are sufficiently readable.

```go
return Config{
  OriginURL: mustParseURL("http://origin.url"),
  ProxyURL:  mustParseURL("http://proxy.url"),
  SSOURL:    mustParseURL("http://sso.url"),
  SSOSecret: "secret",
  // ...
}
```

(and `go fmt` will align the assignments in a pleasing manner.)